File: 프로그래머스/2/42890. 후보키/후보키.java
import java.util.*;

class Solution {
    public int solution(String[][] relation) {
        int n = relation.length;
        int attrNum = relation[0].length;
        List<Integer> candidateKeys = new ArrayList<>();

        // 비트마스크
        for(int bit=1; bit<(1<<attrNum); bit++) {
            Set<String> set = new HashSet<>();
            
            for (int i=0; i<n; i++) {
                StringBuilder sb = new StringBuilder();
                for (int j=0; j<attrNum; j++) {
                    if ((bit&(1<<j)) != 0) { // j 인덱스 포함할 때
                        sb.append(relation[i][j]).append(",");
                    }
                }
                set.add(sb.toString());
            }

            if (set.size() != n) continue;

            // 최소성 검사
            boolean isMinimal = true;
            for (int key : candidateKeys) {
                if ((key & bit) == key) {
                    isMinimal = false;
                    break;
                }
            }
            if (isMinimal) candidateKeys.add(bit);
        }

        return candidateKeys.size();
    }
}

File: 프로그래머스/2/42890. 후보키/후보키.py
def solution(relation):
    answer = 0
    n = len(relation)
    attrLen = len(relation[0])
    candidateKeys = []
    
    for bit in range(1, 1<<attrLen):
        tuple_set = set()
        
        for i in range(n):
            st = ""
            for j in range(attrLen):
                if bit&(1<<j) != 0:
                    st += relation[i][j]
            tuple_set.add(st)
            
        if len(tuple_set) != n:
            continue
        
        isMinimal = True
        for ck in candidateKeys:
            if ck & bit == ck:
                isMinimal = False
                break
        
        if isMinimal:
            candidateKeys.append(bit)
    
    return len(candidateKeys)